### PR TITLE
Menu widget: draw focus on focused item

### PR DESCRIPTION
**Pull request by hius07**
(1) Focus on the last opened book when closing Reader. Closes https://github.com/koreader/koreader/issues/12574.
(2) Focus on the folder when going up to the parent folder.
(3) Focus on a target item (like "Last bookmark" in bookmark list, or "Current page" in fulltext search results list etc.)
Since the Menu widget is affected, the focus may appear in some other (unexpected) places. If it is the case and annoying, the corresponding module to be fixed, after testing in the field.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12583)

### CODE REVIEW

**Review by NiLuJe — CHANGES_REQUESTED** [review]
Some mostly trivial things to tweak for now, but I might want to spend some more time thinking about this. Hopefully this WE, but no guarantees ;).

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12583#-:-O8Hld310wH9VwO2tTdr:bycu3lv)* status: all files reviewed, 3 unresolved discussions (waiting on @hius07)

___
*[`frontend/ui/widget/focusmanager.lua` line 302 at r1](https://reviewable.io/reviews/koreader/koreader/12583#-O8HjrfZEv8UKfioL8YP:-O8HjrfZEv8UKfioL8YQ:b5z2jyd) ([raw file](https://github.com/koreader/koreader/blob/a3aca644dee2d3a10a3ee1d0436bc31ce32cb4b7/frontend/ui/widget/focusmanager.lua#L302)):*
> ```Lua
> -- In some cases, we may only want to send Focus events on non-Touch devices
> FocusManager.FOCUS_ONLY_ON_NT = (Device:hasDPad() and not Device:isTouchDevice()) and 0 or FocusManager.NOT_FOCUS
> FocusManager.FORCED_FOCUS = 3
> ```

Needs a description explaining that it bypassesany and all heuristics and other flags in `moveFocusTo` (given that the default behavior is otherwise to send the unfocus/focus events, which makes this look redundant at first glance).

Also, needs to be set to `4`, flags is intended to be a bitfield. (e.g., (1<<0, 1<<1, 1<<2, 1<<3, ...).
___
*[`frontend/ui/widget/focusmanager.lua` line 323 at r1](https://reviewable.io/reviews/koreader/koreader/12583#-O8HkIsXClt-LhcxonHn:-O8HkIsXClt-LhcxonHo:b-1dxqw4) ([raw file](https://github.com/koreader/koreader/blob/a3aca644dee2d3a10a3ee1d0436bc31ce32cb4b7/frontend/ui/widget/focusmanager.lua#L323)):*
> ```Lua
>         self.selected.y = y
>         -- widget create new layout on update, previous may be removed from new layout.
>         local is_forced = focus_flags == FocusManager.FORCED_FOCUS
> ```

Needs to be tested properly, it's a bitfield, you have no guarantee there won't be another bit set in there.
___
*[`frontend/ui/widget/menu.lua` line 1014 at r1](https://reviewable.io/reviews/koreader/koreader/12583#-O8HlNVJ25yto7Ly7vX0:-O8HlNVJ25yto7Ly7vX1:b-vhery2) ([raw file](https://github.com/koreader/koreader/blob/a3aca644dee2d3a10a3ee1d0436bc31ce32cb4b7/frontend/ui/widget/menu.lua#L1014)):*
> ```Lua
>             -- NOTE: Since this runs automatically on init,
>             --       we use FOCUS_ONLY_ON_NT as we don't want to see the initial underline on Touch devices.
>             self:moveFocusTo(x, y, is_focused and FocusManager.FORCED_FOCUS or FocusManager.FOCUS_ONLY_ON_NT)
> ```

Untested, but I imagine this no longer makes as much sense as before and we now get to see the initial focus on Touch devices regardless?

**Comment by hius07** [comment]
We will see the initial focus only when we pass `itemnumber` or `itemmatch` to Menu:switchItemTable, i.e. when we have a target item.
We will always see the initial focus in the file browser (if there is no focused file, it will be the item No. 1, usually the upper folder).

**Comment by hius07** [comment]
I override the bitfield check for this special case only.

**Comment by NiLuJe** [comment]
> I override the bitfield check for this special case only.

Which was precisely my point ;).

Don't do that. Make the flag value the next bit (i.e., 4), and switch to an actual mask via bit.and to check it ;).

**Comment by Commodore64user** [comment]
> Since the Menu widget is affected, the focus may appear in some other (unexpected) places.

this right here is already making me nervous.

File: frontend/ui/widget/focusmanager.lua [hunk 1]
```diff
@@ -319,7 +321,7 @@ function FocusManager:moveFocusTo(x, y, focus_flags)
         self.selected.x = x
         self.selected.y = y
         -- widget create new layout on update, previous may be removed from new layout.
-        if Device:hasDPad() then
+        if focus_flags == FocusManager.FORCED_FOCUS or Device:hasDPad() then
```
Comment (NiLuJe) [inline frontend/ui/widget/focusmanager.lua:324]: `if bit.band(focus_flags, FocusManager.FORCED_FOCUS) == FocusManager.FORCED_FOCUS`

Comment (hius07) [reply frontend/ui/widget/focusmanager.lua:324]: What is the reason of this complication that gives the same result?

Comment (NiLuJe) [reply frontend/ui/widget/focusmanager.lua:324]: `focus_flags` is a bitfield, any and all of the flags can be OR'ed in it, it's not 1:1.

I'm not entirely sure anything actually *relies* on that right now, but that's how it was designed ;o).

**Comment by NiLuJe** [comment]
> Untested, but I imagine this no longer makes as much sense as before and we now get to see the initial focus on Touch devices regardless?

Pushed a dirty hack to address this issue for FileChooser, but this is indeed a larger issue, as FORCED_FOCUS effectively breaks FOCUS_ONLY_ON_NT ;).

**Comment by NiLuJe** [comment]
So, err, RFC, I guess?

**Comment by hius07** [comment]
Why don't you want to focus the first item in FC? It shows that we have come from the parent folder.

**Comment by hius07** [comment]
So we focus either 
-a book when came from Reader 
-a folder when went up
-the first item (..) when went down

**Comment by NiLuJe** [comment]
> Why don't you want to focus the first item in FC?

The general dislike about that probably came up early on in comments about https://github.com/koreader/koreader/pull/12361

> It shows that we have come from the parent folder.

And where else would we have come from? ;o).

While the feature makes sense the other way around, when going from parent to children, there's only one parent, so it's pretty obvious where you came from, as there's no other choice.

**Comment by hius07** [comment]
FC first item is focused because `itemnumber` is passed:
https://github.com/koreader/koreader/blob/3bc03d9d33076927a5542fbadf3c59bec0a14d32/frontend/ui/widget/filechooser.lua#L513

**Comment by hius07** [comment]
I want to investigate the usage of `path_items`.

**Comment by NiLuJe** [comment]
To recap my train of thought: 

My original gripe was with showing *any* focus on *init*. That happens to be `../` in FC, which lead me to think some more about it, and realize that having this one focused didn't really *ever* make sense (in the context of the PR, not talking about NT).
That also made the stupid PoC hack I pushed very simple and easy, so, two birds, one stone ;).

TL;DR: Yeah, you're very very welcome to find a better solution ;).

**Comment by hius07** [comment]
I do not see the focused first item in FC now.

**Review by NiLuJe — APPROVED** [review]
Will test the latest iteration later, but LGTLM ;).

> Reviewed 1 of 1 files at r3, 1 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12583#-:-O8XAB3D6tZ1U_0DyPiY:b-hiykcd)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @hius07)

File: frontend/ui/widget/menu.lua [hunk 2]
```diff
@@ -995,11 +995,23 @@ end
...
+            select_number = select_number or 1 -- default to select the first item
+            local x, y
+            local nb_cols = self.layout[1] and #self.layout[1] or 1
+            if nb_cols == 1 then
+                x = 1
+                y = select_number
+            else -- mosaic
+                x = select_number % nb_cols
+                y = (select_number - x) / nb_cols + 1
+            end
```
Comment (poire-z) [inline frontend/ui/widget/menu.lua:1010]: Dunno how much I previously had to make Menu aware of the coverbrowser plugin (I think I didn't need anything), feels a bit sad to have to have it know about this case. But ok, it's quite limited.

**Review by poire-z — COMMENTED** [review]
> but LGTLM ;).

Not familiar with the focus stuff, and dunno how much I will hate it :)
But Looks Good To Lazy Me too :)

**Comment by hius07** [comment]
Unwanted focus in ToC:
https://github.com/koreader/koreader/blob/92bcbe06f61c49e2a23431daa77c16ed30298477/frontend/apps/reader/modules/readertoc.lua#L932-L933
We can
-show focus in the file browser only (last book, last folder when going up), or
-set Menu instance proprty `no-focus` in certain widgets.

**Comment by NiLuJe** [comment]
Right now I'd say FM only probably makes the most sense?

**Comment by hius07** [comment]
Finally, focus in the FM only.

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12583)* status: all files reviewed, 1 unresolved discussion

**Comment by mergen3107** [comment]
@hius07 
Works great, thank you!

**Comment by benoit-pierre** [comment]
This does not seem to work as expected: if in the file manager, I enter the third folder, then go back to the parent, said 3rd folder appears as focused, but it actually does not have keyboard focus. Using the down key will focus the 1st item after "../" (all the while keeping the 3rd folder "focused" too). I also disagree that "../" should not be focused on init when entering a folder.

**Comment by NiLuJe** [comment]
> I also disagree that "../" should not be focused on init when entering a folder.

To clarify, my issue with that would be on *touch* devices only. I don't think I tested this one on other devices, so the final approach may have been a bit too heavy-handed ;).

**Comment by poire-z** [comment]
Seeing this only now, I have to say I really don't like it :) it feels like a bug, what's that rectangle ?!.

Moreover, and this is an issue - and not a question of taste I think - : the highlighted text, probably because of the added border, _moves_ (a few pixels south east) when you select it:

image

image

**Comment by mergen3107** [comment]
Is this on emulator? I don't see these boxes in Book information on Kindle

**Comment by poire-z** [comment]
Yes, on the emulator, which has keys/keypad, like NT Kindles.
Fortunately, we don't witness all that on our modern Kobo/Kindles :)
(Which is a reason for my request at https://github.com/koreader/koreader/pull/12579#issuecomment-2388830007 and the next comment - if adding borders for that focus thingie make all things a little uglier, it should be made to only happen on these ugly devices :))

**Comment by hius07** [comment]
Is it related to this PR?
It draws the focus in FM only.

**Comment by poire-z** [comment]
It's KeyValuePage - so probably not, sorry :/
But it must be another crazy Focus/NT related PR :)
I guess that until recently, the focusing only happened after the first move - so we were not seeing it.

**Comment by NiLuJe** [comment]
Welp, apparently clearly not ;p.

Note that the focus affecting tetx layout is an actual issue with a few widgets, we briefly talked about that in the Focus & NT PRs.
